[DOC] Added tutorial about the numpy.ma module.

melissawm opened the pull request:
Hi all,

This is a tutorial with the basics of the `numpy.ma` module, not intended to cover all features and functions available, but to motivate the use of masked arrays for those who don't understand how it can be valuable. 

I tried to keep it simple and not duplicate the content from [this document](https://numpy.org/devdocs/reference/maskedarray.generic.html); I believe some of the content from that could be moved to the tutorial at some point in the future, as the documentation for masked arrays seems to be a bit mixed between auto-generated docs and examples and explanations.

Feedback, suggestions and comments are welcome!

---

Build output here: https://13184-908607-gh.circle-artifacts.com/0/doc/build/html/user/tutorial-ma.html

melissawm committed:
Added tutorial about the numpy.ma module.

rgommers commented on doc/source/user/plot_covid_1.py:
```diff
@@ -0,0 +1,14 @@
+filename = "who_covid_19_sit_rep_time_series.csv"
+
+nbcases = np.genfromtxt(filename, dtype=np.int, delimiter=",", skip_header=7, usecols=range(3, 17))
+dates = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",", max_rows=1, usecols=range(3, 17))
```
This script is failing to run in CI with an ascii decode error. I haven't tried to debug that, however `dtype=np.unicode_` looks a little fishy. Is auto-detection of the dtype failing?

melissawm replied:
Well, if I don't specify the `dtype` dates is an array of `nan`. Is there a more appropriate choice here to specify string data?

eric-wieser commented on doc/source/user/plot_covid_1.5.py:
```diff
@@ -0,0 +1,15 @@
+
+filename = "who_covid_19_sit_rep_time_series.csv"
+
+nbcases = np.genfromtxt(filename, dtype=np.int, delimiter=",", skip_header=7, usecols=range(3, 17))
```
```suggestion
nbcases = np.genfromtxt(filename, dtype=np.int_, delimiter=",", skip_header=7, usecols=range(3, 17))
```

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,364 @@
+new array, this time masking the invalid values::
+
+   >>> from numpy import ma
+   >>> nbcases_ma = ma.masked_values(nbcases, -1)
```
I'd argue this sort of is a problem with the default behavior of genfromtxt, inserting ``-1`` silently is a dangerous default behavior.

Perhaps it would be better to use the `usemask=True` argument to `genfromtxt` instead?

melissawm replied:
I see what you mean, and that would be the best practice, but in this case I chose to do it in parts because of pedagogical reasons. Showing the problem before you solve it is usually more motivating than just giving the right answer from the start :) I can add a note about the `usemask` parameter there.

eric-wieser commented on doc/source/user/plot_covid_4.py:
```diff
@@ -30,6 +30,8 @@
 plt.xticks([0, 7, 13], dates[[0, 7, 13]])
 plt.yticks([0, np.polyval(params, 7), 10000, 17500])
 plt.legend(['Mainland China', 'Cubic estimate', '7 days after start'])
+plt.title("COVID-19 cumulative cases from Jan 21 to Feb 3 2020 - Mainland China\n"\
```
```suggestion
plt.title("COVID-19 cumulative cases from Jan 21 to Feb 3 2020 - Mainland China\n"
```

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,374 @@
+    
+.. plot:: user/plot_covid_1.py
+    :align: center
+    :include-source: 0
```
I'm a little worried that these python files and this rst file are going to get out of sync. @mattip, do you know of a way to get all of the code in the rst, but split it across multiple plots?

melissawm replied:
I changed this document to use the IPython sphinx directive.

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,374 @@
+Included in the :func:`numpy.genfromtxt` function call, we have selected the
+:class:`numpy.dtype` for each subset of the data (either an integer -
+:class:`numpy.int_` - or a string of characters - :class:`numpy.unicode_`). We
+have also used :doc:`slice <basics.indexing>` syntax to select the right
```
I don't see any slice syntax above - is this comment outdated?

melissawm replied:
It was. Thanks for catching that!

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,374 @@
+
+    >>> china_total = china_masked.data
+    array([  278,   309,   574,   835,    10,    10,    17,    22,    23,
+          25,    28, 11821, 14411, 17238])
```
```suggestion
              25,    28, 11821, 14411, 17238])
```

eric-wieser replied:
```suggestion
    array([  278,   309,   574,   835,    10,    10,    17,    22,    23,
              25,    28, 11821, 14411, 17238])
```

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,374 @@
+
+    >>> indices_china
+    (array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
+        17, 18, 19, 20, 21, 22, 23, 25, 26, 27, 28, 29, 31, 33]),)
```
```suggestion
    (array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 25, 26, 27, 28, 29, 31, 33]),)
```

eric-wieser commented on doc/source/user/who_covid_19_sit_rep_time_series.csv:
```diff
@@ -0,0 +1,115 @@
+﻿Province/States,Country/Region,WHO region,1/21/20,1/22/20,1/23/20,1/24/20,1/25/20,1/26/20,1/27/20,1/28/20,1/29/20,1/30/20,1/31/20,2/1/20,2/2/20,2/3/20,2/4/20,2/5/20,2/6/20,2/7/20,2/8/20,2/9/20,2/10/20,2/11/20,2/12/20,2/13/20,2/14/20,2/15/20,2/16/20,2/17/20,2/18/20,2/19/20,2/20/20,2/21/20,2/22/20,2/23/20,2/24/20,2/25/20,2/26/20,2/27/20,2/28/20,2/29/20,3/1/20,3/2/20,3/3/20
```
The docs fail to build because this file starts with a UTF8 byte-order-mark

melissawm replied:
Seems to work now with `utf-8-sig`, thanks!

eric-wieser commented on doc/source/user/who_covid_19_sit_rep_time_series.csv:
```diff
@@ -0,0 +1,115 @@
+﻿Province/States,Country/Region,WHO region,1/21/20,1/22/20,1/23/20,1/24/20,1/25/20,1/26/20,1/27/20,1/28/20,1/29/20,1/30/20,1/31/20,2/1/20,2/2/20,2/3/20,2/4/20,2/5/20,2/6/20,2/7/20,2/8/20,2/9/20,2/10/20,2/11/20,2/12/20,2/13/20,2/14/20,2/15/20,2/16/20,2/17/20,2/18/20,2/19/20,2/20/20,2/21/20,2/22/20,2/23/20,2/24/20,2/25/20,2/26/20,2/27/20,2/28/20,2/29/20,3/1/20,3/2/20,3/3/20
```
```suggestion
Province/States,Country/Region,WHO region,1/21/20,1/22/20,1/23/20,1/24/20,1/25/20,1/26/20,1/27/20,1/28/20,1/29/20,1/30/20,1/31/20,2/1/20,2/2/20,2/3/20,2/4/20,2/5/20,2/6/20,2/7/20,2/8/20,2/9/20,2/10/20,2/11/20,2/12/20,2/13/20,2/14/20,2/15/20,2/16/20,2/17/20,2/18/20,2/19/20,2/20/20,2/21/20,2/22/20,2/23/20,2/24/20,2/25/20,2/26/20,2/27/20,2/28/20,2/29/20,3/1/20,3/2/20,3/3/20

```

melissawm replied:
Hm. Should I change the original file, or add a workaround? Since we're dealing with an official source for this document I wouldn't like to change it for reproducibility reasons. Also, why doesn't it fail on my machine, can you explain that so I can try and figure out a solution? Thanks!

eric-wieser replied:
Using `encoding='utf-8-sig'` may solve this.

melissawm replied:
Added this, thanks. Let's see if this solves the issue.

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,374 @@
+    >>> dates = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
+    ...                       max_rows=1, usecols=range(3, 17))
+    >>> locations = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
+    ...                           skip_header=7, usecols=(0, 1))
```
If possible,
```suggestion
    >>> nbcases = np.genfromtxt(filename, dtype=np.int_, delimiter=",",
    ...                         skip_header=7, usecols=range(3, 17), encoding='utf8')
    >>> dates = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
    ...                       max_rows=1, usecols=range(3, 17), encoding='utf8')
    >>> locations = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
    ...                           skip_header=7, usecols=(0, 1), encoding='utf8')
```
Good practice when loading international data to use character encodings which can hold international names

melissawm replied:
Sure, same as above, it worked here but maybe it's because my whole system is UTF-8 since that's the standard in Brazil

justbennet commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,364 @@
+    >>> import numpy as np
+    >>> filename = "who_covid_19_sit_rep_time_series.csv"
+
+Let's explore the data inside this file for the first 14 days of records. To
```
I think it might be helpful here to provide a description in words of what is in the file and how it is organized.

This is aimed at researchers and people not very experienced with Python, so understanding the problem and the data will be very helpful so that they can follow the reasoning for choosing, and understanding, the programming techniques that follow.

Perhaps something like

> The data file contains data of different types and is organized as follows.
>
> The first row is a header line that (mostly) describes the data in each column that follow in the rows below, and beginning in the fourth column, the header is the date of the observation.
>
> The second through seventh row contain summary data that is of a different type than that which we are going to examine, so we will need to exclude that from the data with which we will work.
>
> The numerical data we wish to work with begins at column 4, row 8, and extends from there to the rightmost column and the lowermost row.

would be relatively short but add to fuller understanding of why you are doing what follows?

melissawm replied:
Good point, thanks for the suggestion! I'll update it.

eric-wieser commented:
Perhaps the solution to avoiding all the repeated `plot.py` files is to make this tutorial a jupyter notebook instead, and use `nbsphinx` to generate the docs?

rgommers commented:
> Perhaps the solution to avoiding all the repeated `plot.py` files is to make this tutorial a jupyter notebook instead, and use `nbsphinx` to generate the docs?

this may be right

> Another option would be to use the [IPython directive](https://ipython.readthedocs.io/en/stable/sphinxext.html#persisting-the-python-session-across-ipython-directive-blocks), which persists variables across blocks.

Interesting, never seen that before. It's says that that directive is in beta; if it's stable enough we could go for it. 

@melissawm what's your preference, notebook or `.. ipython::` directive?

rossbar commented:
An additional note - the `.. ipython` directive is used in the `random` docs, e.g. see [new-or-different](https://github.com/numpy/numpy/blob/master/doc/source/reference/random/new-or-different.rst)

melissawm commented:
I think I'll try the IPython directive, but I'll also add this tutorial as a notebook in our new repo (along with the other ones we already have). So we'll have both options, if that's ok with you all

rgommers commented:
> So we'll have both options, if that's ok with you all

sounds good to me:)

justbennet commented:
Melissa,

Here are the comments I mentioned on the call.

First, might it be useful to distinguish between 'masked arrays' and 'sparse arrays'?

Second, I think that this topic encompasses what we in the old-school statistics world used to refer to as 'user-defined missing values', and they come up often in the social sciences, and most particularly in survey research.  There can be many reasons why data may be missing.  For a given survey question, there may be different reasons for being missing, each of which might be given a different code.

Sometimes characterizing who does not respond is the actual thing of interest.  One common coding was 9, 99, 999 indicated nonresponse; an 8, 98, 998, might code one reason why it was a nonresponse -- e.g., the subject did not respond because they did not know, they refused for privacy, the question was not applicable to their age group, etc.  In one epidemiological study I worked on in the 80s, they wanted to know which types of health data about themselves were known and not know, so distinguishing between 'not known' and 'not applicable' really mattered.

I was thinking that might be a useful and fairly commonly known case where one might want to 'mask' the missing values.

Third, and finally, I think that many people will be less experienced coders.  Many of them will be first inheriting scripts/programs from others, and later in their career they may well be writing them.

I think that adding comments to the actual code for these tutorials might add a bit of length but that would be offset by enhanced comprehensibility by the less experienced.  So, for example, I might suggest

```
# Note we are using skip_header and usecols to read only portions of the data file into
# each variable.
# Read just the dates for columns 3-7 from the first row
dates = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",", max_rows=1, usecols=range(3, 17), encoding="utf8")

# Read the names text names of the geographic locations from the first two columns,
# skipping the first seven rows
locations = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",", skip_header=7, usecols=(0, 1), encoding="utf8")

# Read the numeric data from just the first 14 days
nbcases = np.genfromtxt(filename, dtype=np.int_, delimiter=",", skip_header=7, usecols=range(3, 17), encoding="utf8")
```
That would also make the code a bit more independent of the text, which makes it better review material, maybe?

The use of `genfromtxt` is not the main content, so the inline comments may make it just that much more transparent what those do, especially when combined with the verbal explanation offered earlier.

More experienced programmers want comments to be sparse and more conceptual, whereas I think beginners like having the comments be both a bit more frequent and a bit more detailed.  I liken it to the 'readers' I had when learning German and French -- many words in the main text had short glossary entries at the bottom of the page or in the margin, and new idioms were briefly explained.  Reference to fuller explanations definitions were then give there.  I try to encourage people who code infrequently to be more verbose in their comments and to assume that the person who will read them will be inexperienced and may not know much about the content matter or study yet.

For longer, more complex, more advanced topics, that strategy gets modified so that we aren't interrupting longer procedures with small details so much, but we also expect that those would be read by the more experienced users/coders.

I usually try to get people to do all the data management bits first, then do their statistics and plotting all together.  So, for example, in the `plot_covid_1.5.py`, I would put the `selected_dates` assignment above the first `plot` command.  That leads nicely then into the subsequent versions of the program where the middle section gets more complex an the `plot` portions remain relatively stable.

Will references to the various number `plot_covid*.py` files be added in the text?  It's been my experience that new users like to see the completed version at each 'revision break'.

melissawm commented:
> Perhaps the solution to avoiding all the repeated `plot.py` files is to make this tutorial a jupyter notebook instead, and use `nbsphinx` to generate the docs?
> 
> Another option would be to use the [IPython directive](https://ipython.readthedocs.io/en/stable/sphinxext.html#persisting-the-python-session-across-ipython-directive-blocks), which persists variables across blocks.

Thanks for this, I found this pretty elegant and much simpler (once I got it working)!

melissawm commented:
Hi Bennet, I'll try to answer you here.

> First, might it be useful to distinguish between 'masked arrays' and 'sparse arrays'?

I'm not sure I see what you mean, unless you mean like sparse matrices. I might be biased by my experience but I don't really see that they're connected in this sense. Can you elaborate?
 
> Second, I think that this topic encompasses what we in the old-school statistics world used to refer to as 'user-defined missing values', and they come up often in the social sciences, and most particularly in survey research. There can be many reasons why data may be missing. For a given survey question, there may be different reasons for being missing, each of which might be given a different code.

This is a nice point, I added this observation to the text.
 
> I think that adding comments to the actual code for these tutorials might add a bit of length but that would be offset by enhanced comprehensibility by the less experienced. 

Added, thanks!

> Will references to the various number `plot_covid*.py` files be added in the text? It's been my experience that new users like to see the completed version at each 'revision break'.

These files are not used anymore since I moved to the ipython sphinx directive, I think this is safer and more elegant, but I plan to redo this tutorial as a notebook soon so we'll have a truly executable version for people to download.

Thanks for your comments and suggestions, it's nice to have a fresh view on this.

melissawm commented:
The CI failures are because I haven't integrated the previous docs changes from master here. I'm not sure how to go about this - should I rebase?

charris commented:
Hmm, The PR is mergeable and  should be applying on top of master for testing.

If you want to rebase, the first thing to do is to squash all you commits into one, which will make the job easier. For eleven commits. 
```
$ git rebase -i HEAD~11
```
And follow the instructions, basically replace 'pick' by 'f' in all but the first commit. Then update your master branch from upstream, change back to the PR branch and do
```
$ git rebase master
```
and if there are problems, do `$ git status` to see which files need fixing, then edit the files to fix up any conflicts (sections marked by "<<<") . When you are done with that
```
$ git add <the fixed files>
$ git rebase --continue
$ git push -f origin HEAD
```

If you have any problems, feel free to ask questions.

melissawm commented:
Looks like it worked, thank you @charris for your help!

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+   # Note we are using skip_header and usecols to read only portions of the data
+   # file into each variable.
+   # Read just the dates for columns 3-7 from the first row
+   dates = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
+	                 max_rows=1, usecols=range(3, 17),
+			 encoding="utf-8-sig")
+   # Read the names text names of the geographic locations from the first two
+   # columns, skipping the first seven rows
+   locations = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
+	                     skip_header=7, usecols=(0, 1),
+			     encoding="utf-8-sig")
+   # Read the numeric data from just the first 14 days
+   nbcases = np.genfromtxt(filename, dtype=np.int_, delimiter=",",
+                           skip_header=7, usecols=range(3, 17),
+			   encoding="utf-8-sig")
```
Indents are all over the place here - part of the problem is you've indented the code block 3 spaces and not 4, the other problem is you're mixing tabs and spaces. I'd recommend either:
```suggestion
    dates = np.genfromtxt(
        filename, dtype=np.unicode_, delimiter=",", max_rows=1,
        usecols=range(3, 17), encoding="utf-8-sig")
    # Read the names text names of the geographic locations from the first two
    # columns, skipping the first seven rows
    locations = np.genfromtxt(
        filename, dtype=np.unicode_, delimiter=",", skip_header=7,
        usecols=(0, 1), encoding="utf-8-sig")
    # Read the numeric data from just the first 14 days
    nbcases = np.genfromtxt(
        filename, dtype=np.int_, delimiter=",", skip_header=7,
        usecols=range(3, 17), encoding="utf-8-sig")
```
or
```suggestion
    dates = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
                          max_rows=1, usecols=range(3, 17),
                          encoding="utf-8-sig")
    # Read the names text names of the geographic locations from the first two
    # columns, skipping the first seven rows
    locations = np.genfromtxt(filename, dtype=np.unicode_, delimiter=",",
                              skip_header=7, usecols=(0, 1),
                              encoding="utf-8-sig")
    # Read the numeric data from just the first 14 days
    nbcases = np.genfromtxt(filename, dtype=np.int_, delimiter=",",
                            skip_header=7, usecols=range(3, 17),
                            encoding="utf-8-sig")
```

melissawm replied:
I think this was a copy-paste problem, and tabs got inserted somehow. Sorry!

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+
+.. ipython:: python
+
+    china_total = nbcases[np.where(locations[:, 1] == 'China')].sum(axis=0)
```
```suggestion
    china_total = nbcases[locations[:, 1] == 'China'].sum(axis=0)
```

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+
+.. ipython:: python
+
+   from numpy import ma
+   nbcases_ma = ma.masked_values(nbcases, -1)
```
```suggestion
    from numpy import ma
    nbcases_ma = ma.masked_values(nbcases, -1)
```

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+Macau and "Unspecified" regions of China. Maybe we can remove those from the
+total sum of cases in China, to get a better understanding of the data.
+
+First, we'll identify the indices of locations in China:
```
```suggestion
First, we'll identify the indices of locations in mainland China:
```
Just to avoid political discussions about Taiwan

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+
+.. ipython:: python
+
+    indices_china = np.where((locations[:, 1] == 'China') & 
+                             (locations[:, 0] != 'Hong Kong') & 
+                             (locations[:, 0] != 'Taiwan') &
+                             (locations[:, 0] != 'Macau') &
+                             (locations[:, 0] != 'Unspecified*'))
+
+Now,
+
+.. ipython:: python
+
+    indices_china
```
Possibly clearer to use:
```suggestion
    china_mask = (
        (locations[:, 1] == 'China') & 
        (locations[:, 0] != 'Hong Kong') & 
        (locations[:, 0] != 'Taiwan') &
        (locations[:, 0] != 'Macau') &
        (locations[:, 0] != 'Unspecified*'))

Now,

.. ipython:: python

    china_mask.nonzero()
```

`np.where` seems to be something that matlab users reach for a lot, but in idiomatic numpy it's frequently unecessary.

melissawm replied:
Nice to know, thanks!

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+- A ``fill_value``, a value that may be used to replace the invalid entries
+  in order to return a standard :class:`numpy.ndarray`.
+
+**When can they be useful?**
```
I think you should make these all rST headings so that they can be linked to.

That would look something like:

```suggestion
When can they be useful?
========================
```

melissawm replied:
This was a point made in my other tutorial (tutorial-svd) where I had done all headings and it was requested that I remove them because they were all showing in a global toc somewhere.

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+
+.. ipython:: python
+
+   nbcases_ma
```
```suggestion
    nbcases_ma
```
just to avoid pain dealing with 4 spaces indents in other python code

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+
+   If you are executing the commands above in the IPython shell, it might be
+   necessary to use the command ``plt.show()`` to show the image window. Note
+   also that we use a semicolon at the end of a line to suppress its output, but
+   this is optional.
```
One more option would be to use the `jupyter-execute` directive from `jupyter_sphinx`, which works in almost the same way as `ipython`, but also lets you download a complete script at the end.

melissawm replied:
Nice to know, I'll leave this one as is for now but I'll definitely have a look at that.

eric-wieser commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,377 @@
+
+.. ipython:: python
+	     
+    china_masked = nbcases_ma[np.where(locations[:, 1] == 'China')].sum(axis=0)
```
```suggestion
    china_masked = nbcases_ma[locations[:, 1] == 'China'].sum(axis=0)
```

melissawm committed:
Fixed whitespaces and removed np.where

melissawm commented:
Hi all, just bumping, are there any blockers for this to go forward? Thanks!

rossbar reviewed (approved):
The content LGTM (great, in fact!) - IMO this is a very nice tutorial. Thanks @melissawm 

Note that I didn't take into consideration any style/formatting discussions from e.g. #16244/#16245 - I assume the goal is to get this in and worry about these details in a subsequent step.

melissawm commented on doc/source/user/tutorial-ma.rst:
```diff
@@ -0,0 +1,376 @@
+   also that we use a semicolon at the end of a line to suppress its output, but
+   this is optional.
+
+The graph has a strange shape from January 24th to January 21st. It would be
```
Corrected; it was actually February 1st, don't know where that typo came from!

melissawm committed:
Fixes typos (PR review)

rossbar commented:
There don't seem to be any big blockers here, so I'm going to go ahead and put this in. It will be a lot easier to incrementally handle any proposed changes/improvements in follow-up PRs. Thanks @melissawm for the great tutorial and all of the reviewers!
